File: src/models/verify.ts
import { sha256 } from '@noble/hashes/sha2.js';
import { bytesToHex } from '@noble/hashes/utils.js';
import type { File } from 'expo-file-system';

import { ModelError, type ModelDescriptor } from './types';

/**
 * Verification strategy, and why it is MD5 by default.
 *
 * Measured on an Android emulator against a 128 MB file (see research 0003):
 *
 *   native MD5                 74.5 MB/s   → 4 GB in ~1 min
 *   streamed SHA-256 in JS      1.1 MB/s   → 4 GB in ~61 min
 *
 * `expo-file-system` exposes native hashing for MD5 only, so SHA-256 has to
 * run in JS over a stream that arrives in 1 KB chunks. That is not a tuning
 * problem: hashing alone accounts for ~48 of those 61 minutes.
 *
 * MD5 is the right tool for this specific job despite its reputation. The
 * property being relied on is *second-preimage* resistance — given the
 * publisher's file and its published digest, produce a *different* file with
 * the same digest — for which MD5 has no practical attack. MD5's broken
 * property is *collision* resistance, where an attacker crafts two files
 * together; that only helps an attacker who controls the published digest,
 * i.e. a hostile publisher, which a stronger hash here would not fix either.
 *
 * The remaining gap is closed by a native SHA-256 module (a follow-up task):
 * until then `sha256` in a descriptor is verifiable via `deep: true`, at the
 * cost above.
 */

/**
 * Throws unless this descriptor can actually be verified under `deep`.
 *
 * Called *before* a download starts as well as after it finishes. On-device
 * testing showed why: checking only at the end meant a catalog entry with no
 * MD5 downloaded 800 MB and then reported that it could not be verified —
 * spending a user's bandwidth, and their mobile data, to learn something
 * knowable from the descriptor alone.
 *
 * A size-only match is trivially satisfied by any file of the right length,
 * so passing one would make "verified" mean almost nothing. Catalog entries
 * carry the publisher's SHA-256 but usually no MD5, so this is the path that
 * pushes callers to `deep` until native SHA-256 exists — see research 0003.
 */
export function assertVerifiable(
  descriptor: ModelDescriptor,
  deep: boolean,
): void {
  if (descriptor.md5) return;
  if (deep && descriptor.sha256) return;

  throw new ModelError(
    'verification-unavailable',
    descriptor.id,
    descriptor.sha256
      ? 'This model publishes only a SHA-256 digest, which cannot be checked ' +
          'quickly on device. Enable deep verification — slow, but thorough — ' +
          'before downloading it.'
      : 'This model carries no checksum, so a download cannot be verified.',
  );
}

export type VerifyOptions = {
  /**
   * Additionally verify SHA-256, when the descriptor carries one. Very slow —
   * see the note above. Off by default.
   */
  deep?: boolean;
  onProgress?: (bytesHashed: number) => void;
};

/**
 * SHA-256 of a file, hashed incrementally from its stream.
 *
 * Streamed rather than buffered because these files are gigabytes and
 * `expo-crypto`'s `digest()` takes a whole `BufferSource`. Retained for the
 * `deep` path and for a future native implementation to be checked against.
 */
export async function hashFile(
  file: File,
  onProgress?: (bytesHashed: number) => void,
): Promise<string> {
  const hasher = sha256.create();
  const reader = file.readableStream().getReader();
  let hashed = 0;

  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      if (value) {
        hasher.update(value);
        hashed += value.byteLength;
        onProgress?.(hashed);
      }
    }
  } finally {
    reader.releaseLock();
  }

  return bytesToHex(hasher.digest());
}

/**
 * Verifies a downloaded file against its descriptor.
 *
 * Checks run cheapest-first: size, then native MD5, then (only when asked)
 * SHA-256. Size is effectively free and catches the common failure — a
 * truncated or interrupted download — without spending any hashing time.
 */
export async function verifyFile(
  file: File,
  descriptor: ModelDescriptor,
  options: VerifyOptions = {},
): Promise<void> {
  const wantsDeep = options.deep === true && Boolean(descriptor.sha256);

  assertVerifiable(descriptor, options.deep === true);

  const info = file.info({ md5: true });

  if (info.size !== descriptor.sizeBytes) {
    throw new ModelError(
      'size-mismatch',
      descriptor.id,
      `Expected ${descriptor.sizeBytes} bytes but found ${info.size ?? 0}. ` +
        `The download is incomplete or the source file changed.`,
    );
  }

  if (descriptor.md5) {
    const actualMd5 = info.md5?.toLowerCase();
    if (!actualMd5) {
      throw new ModelError(
        'storage',
        descriptor.id,
        'Could not read the downloaded file to verify it.',
      );
    }

    if (actualMd5 !== descriptor.md5.toLowerCase()) {
      throw new ModelError(
        'checksum-mismatch',
        descriptor.id,
        `Checksum mismatch. Expected MD5 ${descriptor.md5} but computed ` +
          `${actualMd5}. The file is corrupt or was not served by the ` +
          `expected source; it has not been kept.`,
      );
    }
  }

  if (!wantsDeep) return;

  const actualSha = await hashFile(file, options.onProgress);
  if (actualSha !== descriptor.sha256!.toLowerCase()) {
    throw new ModelError(
      'checksum-mismatch',
      descriptor.id,
      `SHA-256 mismatch. Expected ${descriptor.sha256} but computed ` +
        `${actualSha}. The file has not been kept.`,
    );
  }
}

File: apps/desktop/src-tauri/src/models/verify.rs
//! Verifies a downloaded file against its descriptor (task 12.2, mirroring
//! `apps/mobile/src/models/verify.ts`).
//!
//! Checks run cheapest-first: size, then MD5 if present, then SHA-256 — size
//! is effectively free and catches the common failure (a truncated or
//! interrupted download) without spending any hashing time. SHA-256 is
//! preferred because it is the digest model publishers actually publish; an
//! MD5 can only come from a maintainer downloading the file and computing
//! it, which is a strictly weaker claim however fast.

use super::hashing::{md5_file, sha256_file};
use super::types::{ModelDescriptor, ModelError, ModelErrorCode};
use std::path::Path;

/// Throws unless this descriptor can actually be verified. Called *before* a
/// download starts as well as after it finishes — mobile's on-device testing
/// found that checking only at the end meant a checksum-less catalog entry
/// downloaded most of a multi-hundred-MB file before reporting it could
/// never have been verified. A size-only match is trivially satisfied by any
/// file of the right length, so passing one would make "verified" mean
/// almost nothing.
pub fn assert_verifiable(descriptor: &ModelDescriptor) -> Result<(), ModelError> {
    if descriptor.md5.is_some() || descriptor.sha256.is_some() {
        return Ok(());
    }
    Err(ModelError::new(
        ModelErrorCode::VerificationUnavailable,
        &descriptor.id,
        "This model carries no checksum, so a download cannot be verified.",
    ))
}

pub fn verify_file(path: &Path, descriptor: &ModelDescriptor) -> Result<(), ModelError> {
    assert_verifiable(descriptor)?;

    let actual_size = std::fs::metadata(path)
        .map_err(|cause| {
            ModelError::with_cause(
                ModelErrorCode::Storage,
                &descriptor.id,
                "Could not read the downloaded file to verify it.",
                cause,
            )
        })?
        .len();

    if actual_size != descriptor.size_bytes {
        return Err(ModelError::new(
            ModelErrorCode::SizeMismatch,
            &descriptor.id,
            format!(
                "Expected {} bytes but found {actual_size}. The download is incomplete or the source file changed.",
                descriptor.size_bytes,
            ),
        ));
    }

    if let Some(expected) = &descriptor.md5 {
        let actual = md5_file(path).map_err(|cause| {
            ModelError::with_cause(
                ModelErrorCode::Storage,
                &descriptor.id,
                "Could not hash the downloaded file.",
                cause,
            )
        })?;
        if actual.to_lowercase() != expected.to_lowercase() {
            return Err(ModelError::new(
                ModelErrorCode::ChecksumMismatch,
                &descriptor.id,
                format!(
                    "Checksum mismatch. Expected MD5 {expected} but computed {actual}. The file is corrupt or was not served by the expected source; it has not been kept.",
                ),
            ));
        }
    }

    if let Some(expected) = &descriptor.sha256 {
        let actual = sha256_file(path, |_| {}).map_err(|cause| {
            ModelError::with_cause(
                ModelErrorCode::Storage,
                &descriptor.id,
                "Could not hash the downloaded file.",
                cause,
            )
        })?;
        if actual.to_lowercase() != expected.to_lowercase() {
            return Err(ModelError::new(
                ModelErrorCode::ChecksumMismatch,
                &descriptor.id,
                format!("SHA-256 mismatch. Expected {expected} but computed {actual}. The file has not been kept."),
            ));
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use std::path::PathBuf;

    // Well-known test vectors for the literal bytes "hello world" — not
    // invented, not computed circularly from the code under test.
    const CONTENT: &[u8] = b"hello world";
    const SHA256: &str = "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9";
    const MD5: &str = "5eb63bbbe01eeed093cb22bb8f5acdc3";

    fn scratch_file(label: &str, content: &[u8]) -> PathBuf {
        let dir = std::env::temp_dir().join(format!(
            "sovereign-edge-desktop-verify-{label}-{}-{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_nanos(),
        ));
        std::fs::create_dir_all(&dir).expect("could not create scratch dir");
        let path = dir.join("file.bin");
        let mut file = std::fs::File::create(&path).expect("could not create scratch file");
        file.write_all(content)
            .expect("could not write scratch file");
        path
    }

    fn descriptor(size_bytes: u64, md5: Option<&str>, sha256: Option<&str>) -> ModelDescriptor {
        ModelDescriptor {
            id: "scratch-verify-model".to_string(),
            name: "Scratch".to_string(),
            url: "https://example.org/scratch.gguf".to_string(),
            size_bytes,
            md5: md5.map(String::from),
            sha256: sha256.map(String::from),
            quantization: None,
        }
    }

    #[test]
    fn assert_verifiable_rejects_a_descriptor_with_no_checksum_at_all() {
        let d = descriptor(11, None, None);
        let error = assert_verifiable(&d).expect_err("no checksum must be rejected");
        assert_eq!(error.code, ModelErrorCode::VerificationUnavailable);
    }

    #[test]
    fn assert_verifiable_accepts_md5_only() {
        assert!(assert_verifiable(&descriptor(11, Some(MD5), None)).is_ok());
    }

    #[test]
    fn assert_verifiable_accepts_sha256_only() {
        assert!(assert_verifiable(&descriptor(11, None, Some(SHA256))).is_ok());
    }

    #[test]
    fn verify_file_succeeds_with_correct_size_and_sha256() {
        let path = scratch_file("sha-ok", CONTENT);
        let d = descriptor(CONTENT.len() as u64, None, Some(SHA256));
        assert!(verify_file(&path, &d).is_ok());
    }

    #[test]
    fn verify_file_rejects_a_wrong_sha256() {
        let path = scratch_file("sha-bad", CONTENT);
        let d = descriptor(CONTENT.len() as u64, None, Some(&"0".repeat(64)));
        let error = verify_file(&path, &d).expect_err("wrong sha256 must be rejected");
        assert_eq!(error.code, ModelErrorCode::ChecksumMismatch);
    }

    #[test]
    fn verify_file_succeeds_with_correct_size_and_md5() {
        let path = scratch_file("md5-ok", CONTENT);
        let d = descriptor(CONTENT.len() as u64, Some(MD5), None);
        assert!(verify_file(&path, &d).is_ok());
    }

    #[test]
    fn verify_file_rejects_a_wrong_md5() {
        let path = scratch_file("md5-bad", CONTENT);
        let d = descriptor(CONTENT.len() as u64, Some(&"0".repeat(32)), None);
        let error = verify_file(&path, &d).expect_err("wrong md5 must be rejected");
        assert_eq!(error.code, ModelErrorCode::ChecksumMismatch);
    }

    #[test]
    fn verify_file_rejects_a_size_mismatch_before_hashing() {
        let path = scratch_file("size-mismatch", CONTENT);
        let d = descriptor(CONTENT.len() as u64 + 1, None, Some(SHA256));
        let error = verify_file(&path, &d).expect_err("size mismatch must be rejected");
        assert_eq!(error.code, ModelErrorCode::SizeMismatch);
    }

    #[test]
    fn verify_file_reports_storage_error_for_a_missing_file() {
        let missing = std::env::temp_dir().join(format!(
            "sovereign-edge-desktop-verify-missing-{}-{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_nanos(),
        ));
        let d = descriptor(11, None, Some(SHA256));
        let error = verify_file(&missing, &d).expect_err("a missing file must fail");
        assert_eq!(error.code, ModelErrorCode::Storage);
    }
}
